identity/identity.go

```diff
@@ -0,0 +1,169 @@
+package identity
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	logging "github.com/ipfs/go-log"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+var logger = logging.Logger("identity")
+
+// Identity defaults
+const (
+	DefaultConfigCrypto    = crypto.RSA
+	DefaultConfigKeyLength = 2048
+)
+
+// Identity represents identity of a cluster peer for communication,
+// including the Consensus component.
+type Identity struct {
+	ID         peer.ID
+	PrivateKey crypto.PrivKey
+}
+
+// identityJSON represents a Cluster peer identity as it will look when it is
+// saved using JSON.
+type identityJSON struct {
+	ID         string `json:"id"`
+	PrivateKey string `json:"private_key"`
+}
+
+// New generate a public-private keypair and returns a new Identity.
+func New() (*Identity, error) {
+	// pid and private key generation --
+	priv, pub, err := crypto.GenerateKeyPair(
+		DefaultConfigCrypto,
+		DefaultConfigKeyLength)
+	if err != nil {
+		return nil, err
+	}
+	pid, err := peer.IDFromPublicKey(pub)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Identity{
+		ID:         pid,
+		PrivateKey: priv,
+	}, nil
+}
+
+// SaveJSON saves the JSON representation of the Identity to
+// the given path.
+func (ident *Identity) SaveJSON(path string) error {
+	logger.Info("Saving configuration")
+
+	bs, err := ident.ToJSON()
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(path, bs, 0600)
+}
+
+// ToJSON generates a human-friendly version of Identity.
+func (ident *Identity) ToJSON() (raw []byte, err error) {
+	jID, err := ident.toIdentityJSON()
+	if err != nil {
+		return
+	}
+
+	raw, err = json.MarshalIndent(jID, "", "    ")
+	return
+}
+
+func (ident *Identity) toIdentityJSON() (jID *identityJSON, err error) {
+	// Multiaddress String() may panic
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("%s", r)
+		}
+	}()
+
+	jID = &identityJSON{}
+
+	// Private Key
+	pkeyBytes, err := ident.PrivateKey.Bytes()
+	if err != nil {
+		return
+	}
+	pKey := base64.StdEncoding.EncodeToString(pkeyBytes)
+
+	// Set all identity fields
+	jID.ID = ident.ID.Pretty()
+	jID.PrivateKey = pKey
+	return
+}
+
+// LoadJSON receives a raw json-formatted identity and
+// sets the Config fields from it. Note that it should be JSON
+// as generated by ToJSON().
+func LoadJSON(raw []byte) (*Identity, error) {
```

Comment:
We will need env-var support for this too. `CLUSTER_ID` and `CLUSTER_PRIVATEKEY`.

Reply:
Why? This is not something that gets changed often or that you get from somewhere else and put it manually.
I think it's better if users don't create id and private key on their own.

Reply:
Because in many deployment automations, you will have created the id and private key externally and in many of them it is way more convenient to set them with an env var than to create and place a file in a specific location.